config/config.yml

```diff
@@ -131,6 +131,7 @@ config:
             signing_key: "base64 encoded HMAC key or public RSA/ECDSA pem key"
             jwt_header: "Authorization"
             jwt_url_parameter: null
+            jwt_clock_skew_tolerance_seconds: 300
```

Comment:
I searched online for which default values are commonly used , e.g.
https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.tokens.tokenvalidationparameters.defaultclockskew?view=msal-web-dotnet-latest

Reply:
I see the recommended value as 5s with the range being 0-60s.
https://techdocs.akamai.com/api-definitions/docs/json-web-token-jwt-val#:~:text=Clock%20skew%20amount%20specifies%20the,recommended%20default%20value%20is%205.

Reply:
This change seems in line with the RFC guidance [1] and as @AndreaPQ has pointed out Microsoft is using this value.  

I was unable to find counter-examples that indicated smaller time thresholds like 5 seconds, which is originally what came to my mind when considering the clock skew.  In absence of more information, I'd say let's keep it as implemented.

[1]
>  Implementers MAY provide for some small leeway, usually no more than a few minutes,
From https://www.rfc-editor.org/rfc/rfc7519#section-7.2

Reply:
If the current default is 0 then it should remain the same default after the change to avoid changing existing cluster configs to users unknowingly. This change will also need to be accompanied with documentation for the version its released in.

Reply:
Nice find @DarshitChanpura.  @AndreaPQ as you reported this issue, do you know by how much the tokens were expired, would default to 5 seconds mitigated the problem? 

I think 5 seconds is a good default as it is fairly close to the original behavoir.  Since this value was added to the configuration it can be tweaked by cluster operators to find the 'right spot' for their operations

Reply:
I think we should consider the general case and take advantage of  evaluations made by others .
I propose 300 seconds

Reply:
5 minutes is too large IMO. Having a configurable setting should solve the need for when the auth server and the OpenSearch server have clocks out of sync. 

To make sure I understand exactly what the skew will mean.

| OpenSearch Node | AuthServer |
| ------------------ | ----------- |
| 02-24-2023 14:00:00 (1677265200 epoch) | 02-24-2023 14:00:04 (1677265200 epoch - ahead of OpenSearch Node by 4 s) |                   

User logs in successfully to auth server and server gives back a JWT with the following claims:

```
{
  ...
  "nbf": 1677265204,
  "exp": 1677268804 // 1 hr later
}
```

The OpenSearch node receives it immediately, but since the `nbf` from the auth server token is in the future (from the OpenSearch node's perspective) it is not valid yet. (Current time in epoch on the OpenSearch node `1677265200`)

With the skew setting it will offset from the OpenSearch node's perspective and add to the current timestamp (`Instant.now() + <skew>`) when determining its clock's current time in epoch. With a 5 second offset the OpenSearch node would calculate the current time as `1677265205` and the token would be considered valid. Likewise, the skew is used when validating the expiry so it will shorten the length of the session.

Again, I think 5 minutes is too long for a default as a lot of clusters would effectively shorten sessions by 5 minutes if the clocks of the auth server and OpenSearch node are in sync. I believe the default should remain 0 or closer to 0 for a setting like skew.

Reply:
If the Two system are in synch the session Will be longer by 5 minutes, the tollerance works different for nbf and exp

Reply:
@cwperks Could you describe your desired outcome of this pull request?

Reply:
The skew tolerance should be taken into account  when evaluating the expiration date of the token too, see 
https://www.rfc-editor.org/rfc/rfc7523#section-3  point 4

> 4. The JWT MUST contain an "exp" (expiration time) claim that
> limits the time window during which the JWT can be used.  The
> authorization server MUST reject any JWT with an expiration time
> that has passed, subject to allowable clock skew between
> systems.

at Curity they reccomend 30s
https://curity.io/resources/learn/jwt-best-practices/#9-dealing-with-time-based-claims

Reply:
I see, looking inside `JwtUtils` of jose (https://github.com/apache/cxf/blob/main/rt/rs/security/jose-parent/jose/src/main/java/org/apache/cxf/rs/security/jose/jwt/JwtUtils.java#L44-L81) the skew is added to the OpenSearch server's current clock time when determining if the `nbf` claim is valid, but for expiry the skew is added to the `exp` claim from the token when comparing to the OpenSearch server current clock. 

From the curity link:

```
This should be values of a few seconds, and we don't recommend using more than 30 seconds for this purpose, as this would rather indicate problems with the server, not a common clock skew.
```

They mention a maximum of 30s and that greater clock skew is indicative of a bigger problem. wdyt about setting the default value for this setting to a value <30s?

Reply:
Thanks, what do you think @AndreaPQ to setting the default value to 30 seconds?

Reply:
both 300s and 30s are good alternatives; your  team should review what others are doing and make a choice

Reply:
I've updated this PR to use 30s as the default.